From a221580d5fcfc4a05e5434e29db095fb8dbdc6f2 Mon Sep 17 00:00:00 2001
From: Qi Yadong <yadong.qi@intel.com>
Date: Fri, 30 Aug 2019 10:25:59 +0800
Subject: [PATCH 1/5] loader: efi: calculate STAGE1_RT_SIZE according to
 MAX_CPU_NUM

Calculate STAGE1_RT_SIZE according to MAX_CPU_NUM

Change-Id: Ib2379be58f9856f4230d45d87d05b71e4b943ad3
Signed-off-by: Qi Yadong <yadong.qi@intel.com>
---
 loader/stage0/efi/efi_boot_param.c | 4 ++--
 loader/stage0/efi/stage0_asm.h     | 5 ++++-
 loader/stage0/efi/stage0_entry.S   | 2 +-
 3 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/loader/stage0/efi/efi_boot_param.c b/loader/stage0/efi/efi_boot_param.c
index a192e755..e376af23 100644
--- a/loader/stage0/efi/efi_boot_param.c
+++ b/loader/stage0/efi/efi_boot_param.c
@@ -23,7 +23,7 @@ typedef struct {
 
 typedef struct {
 	uint8_t image_load[EVMM_PKG_BIN_SIZE];
-	uint8_t stage1[STAGE1_IMG_SIZE];
+	uint8_t stage1[STAGE1_RT_SIZE];
 	evmm_payload_t payload;
 } memory_layout_t;
 
@@ -100,7 +100,7 @@ evmm_desc_t *boot_params_parse(uint64_t tos_startup_info, uint64_t loader_addr)
 	evmm_desc->evmm_file.runtime_total_size = ((uint64_t)(p_startup_info->vmm_mem_size));
 
 	evmm_desc->stage1_file.runtime_addr = (uint64_t)loader_mem->stage1;
-	evmm_desc->stage1_file.runtime_total_size = STAGE1_IMG_SIZE;
+	evmm_desc->stage1_file.runtime_total_size = STAGE1_RT_SIZE;
 
 	evmm_desc->sipi_ap_wkup_addr = (uint64_t)p_startup_info->sipi_ap_wkup_addr;
 #ifdef LIB_EFI_SERVICES
diff --git a/loader/stage0/efi/stage0_asm.h b/loader/stage0/efi/stage0_asm.h
index 0542bb3f..84744d43 100644
--- a/loader/stage0/efi/stage0_asm.h
+++ b/loader/stage0/efi/stage0_asm.h
@@ -21,7 +21,10 @@
  */
 #define STAGE0_STACK_SIZE             0x4000
 
-#define STAGE1_IMG_SIZE	              0xC000
+/* Temp STACK for stage1 to launch APs */
+#define AP_TEMP_STACK_SIZE            (0x400 * ((MAX_CPU_NUM) + 1))
+
+#define STAGE1_RT_SIZE                (0xA000 + (AP_TEMP_STACK_SIZE))
 #define SEED_MSG_DST_OFFSET           0
 
 /* This payload memory will store evmm_desc_t and device_sec_info_t */
diff --git a/loader/stage0/efi/stage0_entry.S b/loader/stage0/efi/stage0_entry.S
index 91adaf23..0bef3562 100644
--- a/loader/stage0/efi/stage0_entry.S
+++ b/loader/stage0/efi/stage0_entry.S
@@ -38,7 +38,7 @@ tos_image_header:
 	/* entry offset */
 	.long  start_x64 - start
 	/* tos_ldr_size: EVMM_PKG_BIN_SIZE + STAGE1_IMG_SIZE + EVMM_PAYLOAD_SIZE */
-	.long  EVMM_PKG_BIN_SIZE + STAGE1_IMG_SIZE + EVMM_PAYLOAD_SIZE
+	.long  EVMM_PKG_BIN_SIZE + STAGE1_RT_SIZE + EVMM_PAYLOAD_SIZE
 	.byte  TOS_STARTUP_VERSION
 	/* reserved */
 	.fill 3, 1, 0
-- 
2.21.0

